src/solver/linesearch/hagerzhang.rs

```diff
@@ -325,17 +325,17 @@ where
             self.update(op, (a_x, a_f, a_g), (b_x, b_f, b_g), (c_x, c_f, c_g))?;
 
         // S2
-        if (c_x - bb_x).abs() < F::epsilon() {
+        if c_x == bb_x {
```

Comment:
`c_x`, `bb_x`, `aa_x`, and so on are floating point numbers. Floating point numbers should never be compared via `==`, because due to the limited precision, rounding errors will accumulate differently for different mathematical operations. Therefore, even if two numbers should mathematically be the same, they could be different when represented as floating point numbers. Therefore the general approach is to check that the difference between two numbers is small.